File: ios/Packages/GemstonePrimitives/Sources/Device.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import Gemstone

public func generateDeviceKeyPair() -> (privateKey: Data, publicKey: Data) {
    let keyPair = Gemstone.generateDeviceKeyPair()
    return (keyPair.privateKey, keyPair.publicKey)
}

public func devicePublicKey(privateKey: Data) throws -> Data {
    try Gemstone.devicePublicKey(privateKey: privateKey)
}

public func signDeviceAuth(
    privateKey: Data,
    method: String,
    path: String,
    walletId: String,
    body: Data,
    timestampMs: UInt64
) throws -> String {
    try Gemstone.signDeviceAuth(
        privateKey: privateKey,
        method: method,
        path: path,
        walletId: walletId,
        body: body,
        timestampMs: timestampMs
    )
}

File: core/gemstone/src/device.rs
use crate::GemstoneError;

/// Ed25519 device key pair for backend (Gem API) request authentication; the private key is the 32-byte seed.
#[derive(Debug, Clone, uniffi::Record)]
pub struct GemDeviceKeyPair {
    pub private_key: Vec<u8>,
    pub public_key: Vec<u8>,
}

#[uniffi::export]
pub fn generate_device_key_pair() -> GemDeviceKeyPair {
    let seed = rand::random::<[u8; 32]>();
    let public_key = gem_auth::device_public_key(&seed).expect("32 bytes is a valid Ed25519 seed");
    GemDeviceKeyPair {
        private_key: seed.to_vec(),
        public_key: public_key.to_vec(),
    }
}

#[uniffi::export]
pub fn device_public_key(private_key: Vec<u8>) -> Result<Vec<u8>, GemstoneError> {
    Ok(gem_auth::device_public_key(&private_key).map_err(GemstoneError::from)?.to_vec())
}

#[uniffi::export]
pub fn sign_device_auth(private_key: Vec<u8>, method: String, path: String, wallet_id: String, body: Vec<u8>, timestamp_ms: u64) -> Result<String, GemstoneError> {
    gem_auth::build_device_auth_header(&private_key, &method, &path, &wallet_id, &body, timestamp_ms).map_err(GemstoneError::from)
}
